packages/legend-query-builder/src/components/QueryLoaderState.ts

```diff
@@ -0,0 +1,290 @@
+  id: string;
+  owner: string | undefined;
+};
+
+export class QueryLoaderState {
+  readonly applicationStore: GenericLegendApplicationStore;
+  loadQueriesState = ActionState.create();
+  queries: LightQuery[] = [];
+  isQueryLoaderOpen = false;
+  searchText = '';
+  showCurrentUserQueriesOnly = false;
+  showPreviewViewer = false;
+  querypreviewContent?: QueryInfo;
+  searchSpecificationFilters?: QueryTaggedValue[];
+  searchSpecificationProjectCoordinates?: QueryProjectCoordinates[];
+  currentUserId: string | undefined;
+  persistedQueries: QueryPersistInfo[] = [];
+
+  constructor(applicationStore: GenericLegendApplicationStore) {
+    makeObservable(this, {
+      isQueryLoaderOpen: observable,
+      querypreviewContent: observable,
+      queries: observable,
+      showCurrentUserQueriesOnly: observable,
+      showPreviewViewer: observable,
+      searchText: observable,
+      setSearchText: action,
+      setIsQueryLoaderOpen: action,
+      setQueries: action,
+      deleteQuery: action,
+      setShowCurrentUserQueriesOnly: action,
+      setShowPreviewViewer: action,
+      persistQueryIds: action,
+      loadQueries: flow,
+      getPreviewQueryContent: flow,
+      initializeWithCustomQueries: flow,
+      initialize: flow,
+    });
+    this.applicationStore = applicationStore;
+    const persistedQueries =
+      this.applicationStore.settingService.getObjectValue(
+        QUERY_BUILDER_SETTING_KEY.RECENTLY_VIEWED_QUERIES,
+      );
+    this.persistedQueries = persistedQueries
+      ? (persistedQueries as QueryPersistInfo[])
+      : [];
+  }
+
+  setSearchText(val: string): void {
+    this.searchText = val;
+  }
+
+  setIsQueryLoaderOpen(val: boolean): void {
+    this.isQueryLoaderOpen = val;
+  }
+
+  setQueries(val: LightQuery[]): void {
+    this.queries = val;
+  }
+
+  setShowPreviewViewer(val: boolean): void {
+    this.showPreviewViewer = val;
+  }
+
+  setShowCurrentUserQueriesOnly(val: boolean): void {
+    this.showCurrentUserQueriesOnly = val;
+  }
+
+  deleteQuery(query: LightQuery): void {
+    deleteEntry(this.queries, query);
+  }
+
+  persistQueryIds(val: LightQuery | Query): void {
+    let persistQuery = false;
+    if (
+      this.persistedQueries.length < DEFAULT_TYPEAHEAD_SEARCH_LIMIT &&
+      !this.persistedQueries.find((query) => query.id === val.id)
+    ) {
+      persistQuery = true;
+    } else if (this.persistedQueries.find((query) => query.id === val.id)) {
+      persistQuery = true;
+      const idx = this.persistedQueries.findIndex(
+        (query) => query.id === val.id,
+      );
+      this.persistedQueries.splice(idx, 1);
+    }
+    if (persistQuery) {
+      if (val.id === this.currentUserId) {
+        this.persistedQueries.unshift({
+          id: val.id,
+          owner: val.owner,
+        });
+      } else {
+        const idx = this.persistedQueries.findIndex(
+          (q) => q.id !== this.currentUserId,
+        );
+        if (idx < this.persistedQueries.length) {
```

Comment:
confusing to use `idx < this.persistedQueries.length` as a condition for whether that value exists
usually we leverage -1 as a check if the idx is not found for the value else it has been found.